rocks-lib/src/lockfile/mod.rs

```diff
@@ -0,0 +1,199 @@
+use std::io::Write;
+use std::{collections::HashMap, fs::File, io::ErrorKind, path::PathBuf};
+
+use eyre::{bail, Result};
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+use crate::lua_package::{LuaPackage, PackageName, PackageVersion};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub version: PackageVersion,
+    pub pinned: bool,
+    pub dependencies: Vec<String>,
+}
+
+impl LockedPackage {
+    pub fn from(package: &LuaPackage) -> Self {
+        Self {
+            name: package.name().clone(),
+            version: package.version().clone(),
+            pinned: false,
+            dependencies: Vec::default(),
+        }
+    }
+
+    pub fn pinned(mut self, pin: bool) -> Self {
+        self.pinned = pin;
+        self
+    }
+
+    pub fn id(&self) -> String {
+        let mut hasher = Sha256::new();
+
+        hasher.update(format!("{}{}{}", self.name, self.version, self.pinned));
+
+        hex::encode(hasher.finalize())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Lockfile {
+    #[serde(skip_serializing, skip_deserializing)]
+    filepath: PathBuf,
+    // TODO: Serialize this directly into a `Version`
```

Comment:
suggestion: todo comments are easy to forget. Better to create an issue if it's too much for this PR.

Reply:
I always go through my codebase with todo-comments.nvim, so I rarely miss them. But sure, adding an issue sounds like an idea.

Reply:
A benefit of issues is that sometimes an external contributor will open a PR for it.